mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp

```diff
@@ -0,0 +1,466 @@
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto valueTy = op.getValueType();
+    auto tdescTy = op.getTensorDescType();
+    auto shape = tdescTy.getShape();
+
+    auto maybeTargetShape = getTargetShape(op);
+    if (!maybeTargetShape || llvm::equal(*maybeTargetShape, shape))
+      return failure();
+    auto targetShape = *maybeTargetShape;
+
+    auto convertedValTypes = getUnrolledTypes(valueTy, targetShape);
+    auto convertedTdescTypes = getUnrolledTypes(tdescTy, targetShape);
+
+    auto convertedValues =
+        pack(op.getValue(), convertedValTypes, targetShape, loc, rewriter);
+    auto convertedTdescs = pack(op.getTensorDesc(), convertedTdescTypes,
+                                targetShape, loc, rewriter);
+
+    for (auto [v, t] : llvm::zip(convertedValues, convertedTdescs)) {
+      rewriter.create<xegpu::StoreNdOp>(loc, v, t, op.getL1HintAttr(),
+                                        op.getL2HintAttr(), op.getL3HintAttr());
+    }
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+struct UnrollDpasOp : public UnrollPattern<xegpu::DpasOp> {
+  using UnrollPattern<xegpu::DpasOp>::UnrollPattern;
+  LogicalResult matchAndRewrite(xegpu::DpasOp op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+
+    // expecting every operands is a 2D Vector
+    if (llvm::any_of(op->getOperandTypes(), [&](Type type) {
+          auto vecTy = dyn_cast<VectorType>(type);
+          return !vecTy || vecTy.getRank() != 2;
+        }))
+      return failure();
+
+    // a vector of 3 elements should be returned, representing M, K, N
+    // respectively.
+    auto maybeTargetShape = getTargetShape(op);
+    if (!maybeTargetShape || maybeTargetShape->size() != 3)
+      return failure();
+    auto M = (*maybeTargetShape)[0];
+    auto K = (*maybeTargetShape)[1];
+    auto N = (*maybeTargetShape)[2];
+
+    int64_t aBlockSize[2] = {M, K};
+    int64_t bBlockSize[2] = {K, N};
+    int64_t cBlockSize[2] = {M, N};
+
+    auto packWrapper = [&](TypedValue<VectorType> val,
+                           llvm::ArrayRef<int64_t> blockSize) {
+      VectorType type = val.getType();
+      auto maybeGrids = computeShapeRatio(type.getShape(), blockSize);
+      assert(maybeGrids && "Expecting grids to be computed.");
+      auto grids = *maybeGrids;
+      auto numNewOps = computeProduct(grids);
+      if (numNewOps == 1)
+        return llvm::SmallVector<Value>({val});
+      auto newVecTy = type.cloneWith(blockSize, type.getElementType());
+      llvm::SmallVector<Type> convertedTypes(numNewOps, newVecTy);
+      auto values = pack(val, convertedTypes, blockSize, loc, rewriter);
+      return llvm::to_vector(values);
+    };
+
+    auto a = op.getLhs();
+    auto b = op.getRhs();
+    auto c = op.getAcc();
+
+    auto aShape = a.getType().getShape();
+    auto bShape = b.getType().getShape();
```

Comment:
Can these be dynamic shapes?

Reply:
No. the dpas definition doesn't support dynamic shape.